src/main/java/org/ibp/api/rest/name/NameTypeResource.java

```diff
@@ -0,0 +1,82 @@
+import org.generationcp.middleware.api.nametype.GermplasmNameTypeRequestDTO;
+import org.ibp.api.domain.common.PagedResult;
+import org.ibp.api.java.impl.middleware.name.NameTypeService;
+import org.ibp.api.rest.common.PaginatedSearch;
+import org.ibp.api.rest.common.SearchSpec;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.web.PageableDefault;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ResponseBody;
+import springfox.documentation.annotations.ApiIgnore;
+
+import java.util.List;
+
+@Api(value = "Name Type Services")
+@Controller
+public class NameTypeResource {
+
+	@Autowired
+	private NameTypeService nameTypeService;
+
+	@ApiOperation(value = "Create a new name type", notes = "Create a new name type")
+	@RequestMapping(value = "/crops/{cropName}/name-types", method = RequestMethod.POST)
+	@PreAuthorize("hasAnyAuthority('ADMIN')")
+	@ResponseBody
+	public ResponseEntity<Integer> createNameType(@PathVariable final String cropName,
+		@RequestParam(required = false) final String programUUID, @RequestBody final GermplasmNameTypeRequestDTO germplasmNameTypeRequestDTO) {
+		return new ResponseEntity<>(this.nameTypeService.createNameType(germplasmNameTypeRequestDTO), HttpStatus.OK);
+	}
+
+	@ApiOperation(value = "It will retrieve all name types", notes = "It will retrieve all name types")
+	@RequestMapping(value = "/crops/{cropName}/name-types", method = RequestMethod.GET)
+	@PreAuthorize("hasAnyAuthority('ADMIN')")
+	@ApiImplicitParams({
+		@ApiImplicitParam(name = "page", dataType = "integer", paramType = "query",
+			value = "page number. Start at " + PagedResult.DEFAULT_PAGE_NUMBER),
+		@ApiImplicitParam(name = "size", dataType = "integer", paramType = "query",
+			value = "Number of records per page."),
+		@ApiImplicitParam(name = "sort", allowMultiple = true, dataType = "string", paramType = "query",
+			value = "Sorting criteria in the format: property,asc|desc. ")
+	})
+	@ResponseBody
+	public ResponseEntity<List<GermplasmNameTypeDTO>> getNameTypes(@PathVariable final String cropName,
+		@RequestParam(required = false) final String programUUID,
+		@ApiIgnore
+		@PageableDefault(page = PagedResult.DEFAULT_PAGE_NUMBER, size = PagedResult.DEFAULT_PAGE_SIZE) final Pageable pageable){
+		final PagedResult<GermplasmNameTypeDTO> resultPage =
```

Comment:
Refer to GermplasmAuditResource to check how Leo implemented the PaginatedSearch. We should start using that format. It is more elegant and allows us to reuse some code.